lib/rspec/matchers/built_in/raise_error.rb

```diff
@@ -204,6 +208,13 @@ def expecting_specific_exception?
         def raise_message_already_set
           raise "`expect { }.to raise_error(message).with_message(message)` is not valid. The matcher only allows the expected message to be specified once"
         end
+
+        def matching_on_message?(expected_error_or_message)
+          RSpec::Support.is_a_matcher?(expected_error_or_message) &&
+            Array(expected_error_or_message.expected).all? do |expected|
+            [String, Regexp].include?(expected.class)
+          end
+        end
```

Comment:
There's a couple problems with this...

* `expected` is an optional part of the matcher protocol some matchers may not have it.
* A matcher like `start_with('Message start').and end_with('Message end')` isn't going to have a `String` or `Regexp` value for `expected`.
* More generally, this is a decent heuristic but misses many cases where the user may be intending to match on the string.